File: desktop/cross/src/pipe-client.cpp
#include "pipe-client.h"
#include <cstring>

#ifdef _WIN32
#include <windows.h>
#else
#include <sys/socket.h>
#include <sys/un.h>
#include <unistd.h>
#endif

PipeClient::PipeClient(const std::string& pipePath)
    : pipePath_(pipePath) {}

PipeClient::~PipeClient() {
    close();
}

bool PipeClient::connect() {
#ifdef _WIN32
    // Windows: connect to named pipe.
    HANDLE hPipe = CreateFileA(
        pipePath_.c_str(),
        GENERIC_READ | GENERIC_WRITE,
        0, nullptr, OPEN_EXISTING, 0, nullptr);
    if (hPipe == INVALID_HANDLE_VALUE) return false;
    fd_ = (int)(intptr_t)hPipe;
    return true;
#else
    // Unix: connect to domain socket.
    fd_ = socket(AF_UNIX, SOCK_STREAM, 0);
    if (fd_ < 0) return false;

    struct sockaddr_un addr;
    memset(&addr, 0, sizeof(addr));
    addr.sun_family = AF_UNIX;
    if (pipePath_.size() >= sizeof(addr.sun_path)) {
        ::close(fd_);
        fd_ = -1;
        return false;
    }
    strncpy(addr.sun_path, pipePath_.c_str(), sizeof(addr.sun_path) - 1);

    if (::connect(fd_, (struct sockaddr*)&addr, sizeof(addr)) != 0) {
        ::close(fd_);
        fd_ = -1;
        return false;
    }
    return true;
#endif
}

void PipeClient::close() {
    if (readThread_.joinable()) {
        readThread_.detach();
    }
    if (fd_ >= 0) {
#ifdef _WIN32
        CloseHandle((HANDLE)(intptr_t)fd_);
#else
        ::close(fd_);
#endif
        fd_ = -1;
    }
}

bool PipeClient::sendEvent(const HelperEvent& event) {
    auto data = serializeHelperEvent(event);
    return writeFrame(data);
}

bool PipeClient::readMessage(HelperMessage& msg) {
    std::vector<uint8_t> data;
    if (!readFrame(data)) return false;
    msg = parseHelperMessage(data.data(), data.size());
    return true;
}

void PipeClient::startReadLoop(std::function<void(const HelperMessage&)> callback) {
    readThread_ = std::thread([this, callback]() {
        while (fd_ >= 0) {
            HelperMessage msg;
            if (!readMessage(msg)) break;
            callback(msg);
        }
    });
}

bool PipeClient::writeFrame(const std::vector<uint8_t>& data) {
    uint32_t len = uint32_t(data.size());
    uint8_t lenBuf[4];
    lenBuf[0] = uint8_t(len);
    lenBuf[1] = uint8_t(len >> 8);
    lenBuf[2] = uint8_t(len >> 16);
    lenBuf[3] = uint8_t(len >> 24);
    if (!writeAll(lenBuf, 4)) return false;
    if (!data.empty()) {
        if (!writeAll(data.data(), data.size())) return false;
    }
    return true;
}

bool PipeClient::readFrame(std::vector<uint8_t>& data) {
    uint8_t lenBuf[4];
    if (!readExact(lenBuf, 4)) return false;
    uint32_t len = uint32_t(lenBuf[0])
                 | (uint32_t(lenBuf[1]) << 8)
                 | (uint32_t(lenBuf[2]) << 16)
                 | (uint32_t(lenBuf[3]) << 24);
    if (len > 10 * 1024 * 1024) return false;
    data.resize(len);
    if (len > 0) {
        if (!readExact(data.data(), len)) return false;
    }
    return true;
}

bool PipeClient::writeAll(const uint8_t* data, size_t len) {
#ifdef _WIN32
    DWORD written;
    HANDLE h = (HANDLE)(intptr_t)fd_;
    while (len > 0) {
        if (!WriteFile(h, data, (DWORD)len, &written, nullptr)) return false;
        data += written;
        len -= written;
    }
    return true;
#else
    while (len > 0) {
        ssize_t n = write(fd_, data, len);
        if (n <= 0) return false;
        data += n;
        len -= size_t(n);
    }
    return true;
#endif
}

bool PipeClient::readExact(uint8_t* buf, size_t len) {
#ifdef _WIN32
    DWORD nread;
    HANDLE h = (HANDLE)(intptr_t)fd_;
    while (len > 0) {
        if (!ReadFile(h, buf, (DWORD)len, &nread, nullptr)) return false;
        if (nread == 0) return false;
        buf += nread;
        len -= nread;
    }
    return true;
#else
    while (len > 0) {
        ssize_t n = read(fd_, buf, len);
        if (n <= 0) return false;
        buf += size_t(n);
        len -= size_t(n);
    }
    return true;
#endif
}

File: desktop/macos/Sources/SpacewaveHelper/PipeClient.swift
import Foundation

// PipeClient connects to the Go process via Unix domain socket
// using 4-byte LE uint32 length-prefixed framing (framedstream protocol).
class PipeClient {
    private var socket: Int32 = -1
    private let pipePath: String
    var onMessage: ((HelperMessage) -> Void)?

    init(pipePath: String) {
        self.pipePath = pipePath
    }

    func connect() throws {
        socket = Darwin.socket(AF_UNIX, SOCK_STREAM, 0)
        guard socket >= 0 else {
            throw HelperError.socketCreate
        }

        var addr = sockaddr_un()
        addr.sun_family = sa_family_t(AF_UNIX)
        let pathBytes = pipePath.utf8CString
        guard pathBytes.count <= MemoryLayout.size(ofValue: addr.sun_path) else {
            throw HelperError.pathTooLong
        }
        withUnsafeMutablePointer(to: &addr.sun_path) { ptr in
            ptr.withMemoryRebound(to: CChar.self, capacity: pathBytes.count) { dest in
                for i in 0..<pathBytes.count {
                    dest[i] = pathBytes[i]
                }
            }
        }

        let addrLen = socklen_t(MemoryLayout<sockaddr_un>.size)
        let result = withUnsafePointer(to: &addr) { ptr in
            ptr.withMemoryRebound(to: sockaddr.self, capacity: 1) { sockPtr in
                Darwin.connect(socket, sockPtr, addrLen)
            }
        }
        guard result == 0 else {
            throw HelperError.connectFailed(errno)
        }
    }

    func sendEvent(_ event: HelperEvent) throws {
        let data = serializeHelperEvent(event)
        try writeFrame(data)
    }

    func readMessage() throws -> HelperMessage {
        let data = try readFrame()
        return parseHelperMessage(data)
    }

    func startReadLoop() {
        DispatchQueue.global(qos: .userInitiated).async { [weak self] in
            guard let self = self else { return }
            while self.socket >= 0 {
                do {
                    let msg = try self.readMessage()
                    DispatchQueue.main.async {
                        self.onMessage?(msg)
                    }
                } catch {
                    // Connection closed or error, exit read loop.
                    break
                }
            }
        }
    }

    func close() {
        if socket >= 0 {
            Darwin.close(socket)
            socket = -1
        }
    }

    // Write a frame: 4-byte LE uint32 length prefix + data.
    private func writeFrame(_ data: Data) throws {
        var length = UInt32(data.count).littleEndian
        let lenData = Data(bytes: &length, count: 4)
        try writeAll(lenData)
        try writeAll(data)
    }

    // Read a frame: 4-byte LE uint32 length prefix + data.
    private func readFrame() throws -> Data {
        let lenData = try readExact(4)
        let length = lenData.withUnsafeBytes { $0.load(as: UInt32.self) }
        let msgLen = UInt32(littleEndian: length)
        guard msgLen <= 10 * 1024 * 1024 else {
            throw HelperError.messageTooLarge
        }
        return try readExact(Int(msgLen))
    }

    private func writeAll(_ data: Data) throws {
        try data.withUnsafeBytes { ptr in
            var remaining = data.count
            var offset = 0
            while remaining > 0 {
                let written = Darwin.write(socket, ptr.baseAddress!.advanced(by: offset), remaining)
                guard written > 0 else {
                    throw HelperError.writeFailed
                }
                offset += written
                remaining -= written
            }
        }
    }

    private func readExact(_ count: Int) throws -> Data {
        var buffer = Data(count: count)
        var remaining = count
        var offset = 0
        while remaining > 0 {
            let n = buffer.withUnsafeMutableBytes { ptr in
                Darwin.read(socket, ptr.baseAddress!.advanced(by: offset), remaining)
            }
            guard n > 0 else {
                throw HelperError.readFailed
            }
            offset += n
            remaining -= n
        }
        return buffer
    }
}

enum HelperError: Error {
    case socketCreate
    case pathTooLong
    case connectFailed(Int32)
    case messageTooLarge
    case writeFailed
    case readFailed
}
